### PR TITLE
sched/queue/memory: make Dequeue do the right thing

### PR DESCRIPTION
**Pull request by dbentley**

### CODE REVIEW

File: sched/queue/memory/simple.go [hunk 1]
```diff
@@ -3,40 +3,116 @@ package memory
...
 }
 
-func (i simpleWorkItem) Dequeue() {
+type simpleWorkItem struct {
+	job       sched.Job
+	dequeueCh chan struct{}
+}
+
+func (i *simpleWorkItem) Job() sched.Job {
+	return sched.Job(i.job)
```
Comment (jasonbs10) [inline sched/queue/memory/simple.go]: Why cast/construct job, it's already the right type, and you're not doing a deep copy.

Comment (jasonbs10) [inline sched/queue/memory/simple.go]: What is the purpose of this cast/copy?

File: sched/queue/memory/simple.go [hunk 2]
```diff
@@ -3,40 +3,116 @@ package memory
...
+	}
+	select {
+	case outCh <- item:
+		s.dequeueCh = item.dequeueCh
+	case req, ok := <-s.inCh:
+		if !ok {
+			s.inCh = nil
+			return
+		}
+		switch req := req.(type) {
```
Comment (jasonbs10) [inline sched/queue/memory/simple.go]: If we're going to have multiple types, I'd add a comment about it to justify the switch here and the typeless channel.

File: sched/queue/memory/simple.go [hunk 3]
```diff
@@ -3,40 +3,116 @@ package memory
...
+	inCh      chan queueReq
+	outCh     chan queue.WorkItem
+	dequeueCh chan struct{}
+	items     []*simpleWorkItem
+	nextID    int
+	capacity  int
 }
 
-type simpleWorkItem sched.Job
+func (s *simpleQueueState) loop() {
```
Comment (jasonbs10) [inline sched/queue/memory/simple.go]: Do we gain something by not incorporating iter() directly?

File: sched/queue/memory/simple.go [hunk 4]
```diff
@@ -3,40 +3,116 @@ package memory
...
+	job       sched.Job
+	dequeueCh chan struct{}
+}
+
+func (i *simpleWorkItem) Job() sched.Job {
+	return sched.Job(i.job)
+}
+
+func (i *simpleWorkItem) Dequeue() {
+	i.dequeueCh <- struct{}{}
```
Comment (jasonbs10) [inline sched/queue/memory/simple.go]: Is this type a placeholder for something? Or is this considered void (I've been using bool in its place).

**Comment by jasonbs10** [comment]
lgtm once the build succeeds.

File: sched/queue/memory/simple.go (hunk 1 above)
Comment (dbentley) [inline sched/queue/memory/simple.go]: Fixed, thanks. (It was a bad merge thing)

File: sched/queue/memory/simple.go (hunk 2 above)
Comment (dbentley) [inline sched/queue/memory/simple.go]: Removed. Thanks.

File: sched/queue/memory/simple.go (hunk 3 above)
Comment (dbentley) [inline sched/queue/memory/simple.go]: Saving a level of indentation, which used to be an issue, but isn't. Fixed.

File: sched/queue/memory/simple.go (hunk 4 above)
Comment (dbentley) [inline sched/queue/memory/simple.go]: It's void. I prefer it to bool because with bool you have to think "wait, does the value matter?"

Comment (dbentley) [inline sched/queue/memory/simple.go]: WDYT?

Comment (jasonbs10) [inline sched/queue/memory/simple.go]: That's fine. You only have to think about the void once, you have to think about the bool every time you encounter one.

**Pull request merged**